File: ESTRUCTURA DE DATOS/Proyecto Final/SimuladorDeFirewallBasico.py
import ipaddress

def EvaluarPaquete(ip: str, puerto: int, protocolo: str) -> bool:

    permiso = False  #Todo esta bloqueado por defecto (Implicit Deny)

    #Regla 1: Permitir DNS (UDP puerto 53)
    if protocolo.upper() == 'UDP' and puerto == 53:
        permiso = True

    #Regla 2: Permitir ping (ICMP) desde la red local "192.168.x.x"
    if ip.startswith("192.168") and protocolo.upper() == 'ICMP':
        permiso = True

    #Regla 3: Bloqueo explicito de RDP (TCP 3389)
    if protocolo.upper() == 'TCP' and puerto == 3389:
        permiso = False

    return permiso


def main():
    print("===== Firewall INFOCORP =====")

    #Validar IP
    while True:
        ip = input("Ingrese la IP de origen: ").strip()
        try:
            ipaddress.ip_address(ip)
            break  # Si es válida, salir del bucle
        except ValueError:
            print(f"❌ Error: '{ip}' no es una IP válida. Intente de nuevo.")

    #Validar puerto
    while True:
        try:
            puerto = int(input("Ingrese el puerto de destino: ").strip())
            if 0 <= puerto <= 65535:
                break
            else:
                print("❌ Error: el puerto debe estar entre 0 y 65535.")
        except ValueError:
            print("❌ Error: el puerto debe ser un número entero.")

    #Validar protocolo
    protocolos_validos = {"TCP", "UDP", "ICMP"}
    while True:
        protocolo = input("Ingrese el protocolo (TCP, UDP, ICMP): ").strip().upper()
        if protocolo in protocolos_validos:
            break
        else:
            print("❌ Error: protocolo no válido. Debe ser TCP, UDP o ICMP.")

    #Evaluacion del paquete
    resultado = EvaluarPaquete(ip, puerto, protocolo)

    #Resultado
    if resultado:
        print("✅ Paquete permitido")
    else:
        print("⛔ Paquete bloqueado")


#Ejecutar el programa principal
if __name__ == "__main__":
    main()

File: C++/Trabajo Final C++/SimuladorDeFirewallBasico.cpp
// Simulador de Firewall Basico - Dennis Montero

#include <iostream>
#include <string>
#include <sstream>
#include <cctype>

using namespace std;

// Funcion para validar IPv4
bool validarIP(const string& ip) {
    int num;
    char punto;
    stringstream ss(ip);

    for (int i = 0; i < 4; i++) {
        if (!(ss >> num)) return false;
        if (num < 0 || num > 255) return false;
        if (i < 3 && !(ss >> punto && punto == '.')) return false;
    }
    return ss.eof();
}

// Convertir string a mayusculas
string aMayusculas(string texto) {
    for (char &c : texto)
        c = toupper(c);
    return texto;
}

// Evaluacion del paquete segun reglas del firewall
bool EvaluarPaquete(string ip, int puerto, string protocolo) {

    bool permiso = false; // Implicit Deny

    // Regla 1: Permitir DNS (UDP 53)
    if (protocolo == "UDP" && puerto == 53)
        permiso = true;

    // Regla 2: Permitir ICMP desde 192.168.x.x
    if (ip.rfind("192.168", 0) == 0 && protocolo == "ICMP")
        permiso = true;

    // Regla 3: Bloqueo explícito RDP (TCP 3389)
    if (protocolo == "TCP" && puerto == 3389)
        permiso = false;

    return permiso;
}

int main() {

    string ip, protocolo;
    int puerto;

    cout << "===== Firewall INFOCORP =====" << endl;

    // Validar IP
    while (true) {
        cout << "Ingrese la IP de origen: ";
        cin >> ip;

        if (validarIP(ip))
            break;
        else
            cout << "Error: IP no valida. Intente de nuevo.\n";
    }

    // Validar puerto
    while (true) {
        cout << "Ingrese el puerto de destino: ";
        cin >> puerto;

        if (cin.fail() || puerto < 0 || puerto > 65535) {
            cout << "Error: puerto invalido.\n";
            cin.clear();
            cin.ignore(1000, '\n');
        } else
            break;
    }

    // Validar protocolo
    while (true) {
        cout << "Ingrese el protocolo (TCP, UDP, ICMP): ";
        cin >> protocolo;

        protocolo = aMayusculas(protocolo);

        if (protocolo == "TCP" || protocolo == "UDP" || protocolo == "ICMP")
            break;
        else
            cout << "Error: protocolo no valido.\n";
    }

    // Evaluacion del paquete
    bool resultado = EvaluarPaquete(ip, puerto, protocolo);

    // Resultado
    cout << "===== Resultado del analisis =====" << endl;
    if (resultado)
        cout << "Paquete permitido\n";
    else
        cout << "Paquete bloqueado\n";

    return 0;
}
